<a href="https://colab.research.google.com/github/pandeyp84/Descriptive_Statistics_using_Python/blob/main/Transformers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Submitted by : Prabhakar.Pandey@gmail.com

The transformers architecture comprises encoder and decoder models that work together to generate insightful results.

**Pipeline API**

The pipeline API is comprised of a high-level API that performs all the required steps, including preprocessing of inputs and so on. It supports a variety of NLP tasks, including Sentiment Analysis, Feature Extraction, Question Answering, Summarization.

In [1]:
!pip install transformers[sentencepiece]

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 46.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.2/311.2 kB 39.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 101.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 89.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 79.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 40.7 MB/s eta 0:00:00


**Sentiment Analysis**

In [2]:
from transformers import pipeline


In [3]:
classifier = pipeline('sentiment-analysis')

No model was supplied, defaulted to distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


The pipeline API chooses a fine-tuned, pre-trained model for the purpose of sentiment analysis in English. We downloaded this model and cached it after creating the classifier object. Therefore, after rerunning the above command, the cached model will be loaded instead of being downloaded.

We can verify which model the present classifier object is using in the following way:

In [4]:
classifier.model.name_or_path

'distilbert-base-uncased-finetuned-sst-2-english'

Here, the classifier chooses a type of BERT model.

In [6]:
classifier("I am  really excited about about today !! ")

[{'label': 'POSITIVE', 'score': 0.9997952580451965}]

We observe that the classifier has classified the input sentence as a POSITIVE sentence with almost 100% confidence.

In [7]:
classifier(["I am very excited for this new movie !!",
            "I not very unhappy",
            "I hate this weather !!",
            "I really hate that movie.."])

[{'label': 'POSITIVE', 'score': 0.9998107552528381},
 {'label': 'POSITIVE', 'score': 0.8951895236968994},
 {'label': 'NEGATIVE', 'score': 0.9992252588272095},
 {'label': 'NEGATIVE', 'score': 0.9994947910308838}]

With a few more examples, we have verified that the classifier has classified the sentiment of the sentences correctly with good confidence.

**Zero-Shot Classification**

In Zero-Shot classification, the input texts are not labeled. Here, we need to define the labels as per our needs.

In [8]:
from transformers import pipeline
classifier = pipeline('zero-shot-classification')

No model was supplied, defaulted to facebook/bart-large-mnli and revision c626438 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


In [9]:
classifier("This is really a very good course about how to cook mutton",
           candidate_labels = ["Education", "Eat", "Cooking"])

{'sequence': 'This is really a very good course about how to cook mutton',
 'labels': ['Cooking', 'Eat', 'Education'],
 'scores': [0.7549446225166321, 0.226561039686203, 0.018494367599487305]}

It is almost 75% confident that the sentence is about cooking.

In [10]:
classifier(
    ["This is a course on advanced deep learning",
     "This AI course can generate more than 10 million in revenue",],
    candidate_labels=["education", "ai", "business"],
)

[{'sequence': 'This is a course on advanced deep learning',
  'labels': ['ai', 'education', 'business'],
  'scores': [0.721426248550415, 0.21009023487567902, 0.06848350167274475]},
 {'sequence': 'This AI course can generate more than 10 million in revenue',
  'labels': ['ai', 'business', 'education'],
  'scores': [0.6334286332130432, 0.34118345379829407, 0.0253879576921463]}]

It is almost 72% confident that the first sentence is about AI and almost 63% sure that the second sentence is about AI.

**Text Generation**

The text generation is done using the initial prompt, and the model auto-completes the remaining text. However, text generation involves some randomness, and the results may not match exactly.

In [11]:
from transformers import pipeline

generator = pipeline('text-generation')

No model was supplied, defaulted to gpt2 and revision 6c0e608 (https://huggingface.co/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


In [12]:
results = generator("I am really happy because ",
          num_return_sequences=2,
          max_length=30)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


In [14]:
for i in results:
    print(i['generated_text'])
    print('\n')

I am really happy because  I saw how many people were reading the thread and they are all really happy!
The reason my question is so


I am really happy because  I realized the hard to find game here.  It has everything I expected.  I found it even




**Question Answering**

The question answering pipeline can answer questions by understanding the context of the given information.

In [15]:
from transformers import pipeline

question_answerer = pipeline("question-answering")

No model was supplied, defaulted to distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


In [16]:
question_answerer(
    question = "What is the capital of India ?",
    context = """India, officially the Republic of India, is a country in South Asia.
    It is the second-most populous country, the seventh-largest country by land area, and the most populous democracy in the world.
    The capital is New Delhi"""
)

{'score': 0.9928799867630005, 'start': 220, 'end': 229, 'answer': 'New Delhi'}

**Summarization**

The summarization pipeline API has the ability to generate a summary of the given input text by keeping most of the important aspects.

In [18]:
from transformers import pipeline

summarizer = pipeline('summarization')

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


We have to enter the input text along with a max_length or min_length argument for the summary generation of the text.

In [21]:
summarizer("""
America has changed dramatically during recent years. Not only has the number of
    graduates in traditional engineering disciplines such as mechanical, civil,
    electrical, chemical, and aeronautical engineering declined, but in most of
    the premier American universities engineering curricula now concentrate on
    and encourage largely the study of engineering science. As a result, there
    are declining offerings in engineering subjects dealing with infrastructure,
    the environment, and related issues, and greater concentration on high
    technology subjects, largely supporting increasingly complex scientific
    developments. While the latter is important, it should not be at the expense
    of more traditional engineering.

    Rapidly developing economies such as China and India, as well as other
    industrial countries in Europe and Asia, continue to encourage and advance
    the teaching of engineering. Both China and India, respectively, graduate
    six and eight times as many traditional engineers as does the United States.
    Other industrial countries at minimum maintain their output, while America
    suffers an increasingly serious decline in the number of engineering graduates
    and a lack of well-educated engineers.
""", max_length=40)

Your min_length=56 must be inferior than your max_length=40.


[{'summary_text': ' The number of U.S. engineering graduates in traditional engineering disciplines has declined in recent years . China and India graduate six and eight times as many traditional engineers as does the United States .'}]

We can only enter the input text and omit the max_length or min_length argument for the summary generation of the text.

In [22]:
summarizer("""
America has changed dramatically during recent years. Not only has the number of
    graduates in traditional engineering disciplines such as mechanical, civil,
    electrical, chemical, and aeronautical engineering declined, but in most of
    the premier American universities engineering curricula now concentrate on
    and encourage largely the study of engineering science. As a result, there
    are declining offerings in engineering subjects dealing with infrastructure,
    the environment, and related issues, and greater concentration on high
    technology subjects, largely supporting increasingly complex scientific
    developments. While the latter is important, it should not be at the expense
    of more traditional engineering.
    Rapidly developing economies such as China and India, as well as other
    industrial countries in Europe and Asia, continue to encourage and advance
    the teaching of engineering. Both China and India, respectively, graduate
    six and eight times as many traditional engineers as does the United States.
    Other industrial countries at minimum maintain their output, while America
    suffers an increasingly serious decline in the number of engineering graduates
    and a lack of well-educated engineers.
""")

[{'summary_text': ' The number of U.S. engineering graduates in traditional engineering disciplines has declined in recent years . China and India graduate six and eight times as many traditional engineers as does the United States . Rapidly developing economies such as India, Europe and Asia, continue to encourage and advance the teaching of engineering .'}]

**From the pipeline, use any model.**

Let us implement the text generation pipeline object by using the GPT2 model.

In [23]:
from transformers import pipeline

generator = pipeline('text-generation', model='gpt2')

We can verify the model in the following way:

In [24]:
generator.model.name_or_path

'gpt2'

From the above-mentioned examples, it is quite clear that pre-trained models can implement remarkable tasks in the NLP domain.

**Tokenizers**

Like other neural networks, the transformers cannot process the raw input text directly. Preprocessing is required, and this process is called tokenization, where text input is converted into numbers. The preprocessing and tokenization processes are required to be done in the same way during the training of the model. As we are using pre-trained models, the corresponding tokenizer for the model should be used. This can be achieved with the help of the AutoTokenizer class.

To do this, the tokenizer takes the following steps:

(i) The tokenizer splits the input text into words, subwords, or individual letters, which are known as tokens.

(ii) Each token is mapped to a unique integer.

(iii) Arranging and adding required inputs that are impactful to the model.

In [25]:
!pip install transformers[sentencepiece] --q

In [26]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

Here, the AutoTokenizer class has been imported from the transformers library and initialized with the model checkpoint name.

In [27]:
tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path=checkpoint)

Now, we have initialized the tokenizer.

After the execution of the code, we downloaded the tokenizer of the model named distilbert-base-uncased-finetuned-sst-2-english and cached it for further usage.

In [28]:
raw_inputs = [
    "I have been waiting for this HAAI course for last 3 months.",
    "I am very excited about learning new AI models !!",
    "I hate this hot weather which makes me feel irritated  !"
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)


{'input_ids': tensor([[  101,  1045,  2031,  2042,  3403,  2005,  2023,  5292,  4886,  2607,
          2005,  2197,  1017,  2706,  1012,   102],
        [  101,  1045,  2572,  2200,  7568,  2055,  4083,  2047,  9932,  4275,
           999,   999,   102,     0,     0,     0],
        [  101,  1045,  5223,  2023,  2980,  4633,  2029,  3084,  2033,  2514,
         15560,   999,   102,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0]])}


Transformers only accept tensors, and that's why raw texts are first converted into tensors.

After checking the output, we can see input_ids consist of tokenized inputs and attention_mask contains the 1s and 0s which are used by the attention layer in the transformer defining whether the corresponding token should be attended or not.

**Model**

In [29]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

The pre-trained model has been downloaded the same way as the tokenizer.

For loading the pre-trained model, the same from_pretrained method from the AutoModel class has been used.

In [30]:
model = AutoModel.from_pretrained(checkpoint)

In [31]:
outputs = model(**inputs)
print(outputs.last_hidden_state.shape)

torch.Size([3, 16, 768])


We have the outputs with the help of tokenized inputs into the model. From the shape of the output’s hidden state, we observe that our model has a batch size (the number of sequences processed at a time) of 3, a sequence length (the length of each sequence) of 16, and a hidden state (the vector dimension of model input) of size 768.

**Sequence Classification**

In [32]:
import torch
from transformers import AutoModelForSequenceClassification, AutoTokenizer
checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

In [33]:
tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path=checkpoint)

Here we are downloading the model to get some predictions.

In [42]:
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

In [43]:
raw_inputs = [
    "I get irritated during the winter.",
    "Ravi received the director's gold medal for being the topper.",
    "As expected, Sumana received her promotion letter today."
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)


{'input_ids': tensor([[  101,  1045,  2131, 15560,  2076,  1996,  3467,  1012,   102,     0,
             0,     0,     0,     0,     0,     0],
        [  101, 16806,  2363,  1996,  2472,  1005,  1055,  2751,  3101,  2005,
          2108,  1996,  2327,  4842,  1012,   102],
        [  101,  2004,  3517,  1010,  7680,  5162,  2363,  2014,  4712,  3661,
          2651,  1012,   102,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0]])}


Our tokenized inputs are ready. Let us go for the predictions.

In [44]:
outputs = model(**inputs)
print(outputs)

SequenceClassifierOutput(loss=None, logits=tensor([[ 4.0595, -3.3044],
        [-3.9833,  4.3278],
        [-2.4997,  2.6847]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)


In [45]:
print(outputs.logits.shape)

torch.Size([3, 2])


Our model has a matrix dimension of 3 x 2, as we have 3 sequences in the input and there are 2 classes.

We are now finding the outputs by passing them through the softmax activation to get the probabilities of each class for the input sentences.

In [46]:
#Find the label / class probabilities
import torch
outputs = torch.nn.functional.softmax(outputs.logits, dim = -1)
print(outputs)

tensor([[9.9937e-01, 6.3328e-04],
        [2.4571e-04, 9.9975e-01],
        [5.5720e-03, 9.9443e-01]], grad_fn=<SoftmaxBackward0>)


The first input has a probability distribution of [0.99937, 0.00063328], which indicates the model is 99.937% confident that the input is NEGATIVE.

The second input has probabilities [0.00024571, 0.999754], showing the model is 99.9754% confident that the input is POSITIVE.

The third input's probabilities are [0.005572, 0.99443], meaning the model is 99.443% confident that the input is POSITIVE.

We observe [0.99937, 0.00063328] as the output for the first input, indicating a 99.937% confidence level that the input sample belongs to the NEGATIVE class. The second input yields [0.00024571, 0.999754], which corresponds to a 99.9754% confidence level for the POSITIVE class. Finally, the third input shows [0.005572, 0.99443], resulting in a 99.443% confidence level for the POSITIVE class. These results demonstrate the model's high confidence in its predictions for each class.

We can check the labels of the model in the following way:

In [47]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}